### ISSUE TITLE
Wrong type when passing arguments with tuple expansion

### ISSUE THREAD

**Issue by lovetox** [microsoft/pyright#2658]
**Describe the bug**
pyright seems to get types wrong when passing arguments to a function via tuple expansion

**Expected behavior**
Don’t raise reportGeneralTypeIssues

**Screenshots or Code**
```python
from typing import Optional

def test(path: str, something: Optional[str] = None) -> str:
    return path

a = [
    ('a'),
    ('b', 'c')
]

for x in a:
    test(*a)
```

**VS Code extension or command-line**
Running in SublimeText via LSP-Pyright

**Additional context**
pyright version: 1.1.192

**Comment by erictraut** [comment]
This behavior is expected given the inference rules that pyright uses. The inferred type of `a` is `list[str | tuple[str, str]]`. 

Mypy also emits an error in the same case.

You'll need to simplify the code or provide additional type annotations to make this type check without errors.

**Comment by lovetox** [comment]
Sorry my initial example had a typo when iterating the loop thats why it does not work

Please check this out:

```python
from typing import Optional
from pathlib import Path

class A:
    def add(self,
            name: str,
            path: Path,
            location: Optional[str] = None) -> None:
        pass

    def make(self):
        user_dir_paths = [
            ('b', Path('somepath')),
            ('b', Path('somepath'), 'a'),
            ('b', Path('somepath'), 'a'),
            ('b', Path('somepath'), 'a'),
        ]

        for path in user_dir_paths:
            self.add(*path)


class B:
    def add(self,
            name: str,
            path: str,
            location: Optional[str] = None) -> None:
        pass

    def make(self):
        user_dir_paths = [
            ('b', 'path'),
            ('b', 'path', 'a'),
            ('b', 'path', 'a'),
            ('b', 'path', 'a'),
        ]

        for path in user_dir_paths:
            self.add(*path)
```

A fails, B not, only difference is the type of the `path` arg

**Comment by erictraut** [comment]
This code is too clever for static type checking. You'll need to simplify it if you want it to type check in strict modes without errors.

Pyright already contains a bunch of complex special-case logic to handle unpacking of known-length tuples within a call expression. However, it cannot handle unions of tuples that are of different lengths. That's what's happening here.

This code is already sufficiently complex and difficult to maintain, so I'm not inclined to handle further special cases on top of the existing special-case logic. This isn't a common enough use case to justify the added complexity and analyzer performance impact that it would impose.

One very simple change you can make to your code is to include a `None` in the first entry of the `use_dir_paths` entry. That is, change `("b", Path("somepath"))` to `("b", Path("somepath"), None)`. That way, all tuples in the list are of the same length.

Another workaround is to disable strict-mode type checking (or at least the strictListInference rule) for this file.

**Comment by lovetox** [comment]
Ok thanks for the answer, i can make this work.
